cli/src/cli_util.rs

```diff
@@ -1093,17 +1093,37 @@ impl WorkspaceCommandHelper {
         if values.is_empty() {
             Ok(FilesetExpression::all())
         } else if self.settings().config().get_bool("ui.allow-filesets")? {
-            self.parse_union_filesets(ui, values)
+            self.parse_union_filesets(ui, values).or_else(|e| {
+                // If the arguments could not be parsed as valid filesets but could still be
+                // valid paths, warn the user about it.
+                if self.parse_file_path_expressions(values).is_ok() {
```

Comment:
I don't think parsing as a file path would help reduce false positives. Many fileset expressions are valid file paths.

Instead, maybe we can just attach hint to `FilesetParseError` unconditionally (in command_error.rs)?

Reply:
Using a hint is a good idea, I've switched the code to use them, thanks for the suggestion.

For the rest: attaching to the hint in `command_error.rs` without checking that the arguments look like valid file paths is risky. For example, using "../foobar" as a fileset when we are at the root of the directory gives a `FilesetParserError` but it makes no sense to attach this hint, as the problem is that the file is out of the repository. This is why checking that file paths look valid using `self.parse_file_path_expressions(values)` is important to reduce the number of false positive for attaching this hint.

With this change, we get:

```shell
$ jj file show ../foobar
Error: Failed to parse fileset: Invalid file pattern
Caused by:
1:  --> 1:1
  |
1 | ../foobar
  | ^-------^
  |
  = Invalid file pattern
2: Path "../foobar" is not in the repo "."
3: Invalid component ".." in repo-relative path "../foobar"
```

If I attach the hint inconditionally, I get:

```shell
$jj file show ../foobar
Error: Failed to parse fileset: Invalid file pattern
Caused by:
1:  --> 1:1
  |
1 | ../foobar
  | ^-------^
  |
  = Invalid file pattern
2: Path "../foobar" is not in the repo "."
3: Invalid component ".." in repo-relative path "../foobar"
Hint: Filesets are enabled by default. If you want to match file paths containing characters used in fileset expressions (':', or '~'), escape them, or prefix them with "glob:".
```

Reply:
Ah, yes. Perhaps, the hint should be added only to `SyntaxError`?

fwiw, there are more metacharacters in addition to `:` and `~`. It might be better to show a doc link https://martinvonz.github.io/jj/latest/filesets/ instead.

Reply:
I've updated the message to point to the documentation.

Also @yuja, when you mean attaching the hint to `SyntaxError`, would you prefer the implementation in https://github.com/martinvonz/jj/commit/dd70bd34193bf7b9a1eaef40e98dfd76968ef2b8?

Reply:
I meant the hint can be added by `command_error::fileset_parse_error_hint()` or `impl From<FilesetParseError> for CommandError`. I assume it wouldn't be important for user to know whether the fileset is enabled by default or not.

Maybe we can also add a similar hint to revset, to mention that a tag or branch name might need quoting, see this URL (or `jj help revset` or something) for the syntax...

Reply:
I see. It makes the code much more compact indeed. I haven't touched revsets though, this would have to be done in another PR.